### ISSUE TITLE
Feature: Chat with pdf/document

### ISSUE THREAD

**Issue by TrendingApp** [Vali-98/ChatterUI#41]
Could you add a feature where you can chat with a pdf/document and ask  questions related to the pdf/document

**Comment by Vali-98** [comment]
This is a rather complex feature due to several factors:

1. Document parsing - currently unimplemented
2. RAG - not impossible but does incur some level of calculation + difficulty in storage and searching

I do want this feature too, but it will probably be far off.

**Comment by TrendingApp** [comment]
Could you make it work with a text file instead of a pdf?

**Comment by Vali-98** [comment]
No, this will still be a significant amount of work and will not be added soon.

**Comment by softpedia-inc** [comment]
> No, this will still be a significant amount of work and will not be added soon.

When can we expect this feature sir?

**Comment by Sphyker789** [comment]
> No, this will still be a significant amount of work and will not be added soon.

Any updates?

**Comment by lrq3000** [comment]
Just some ideas after trying manually to do simple RAG like queries:

* For a text file/md file, you can just prepend the document's content in the message, and add the user's prompt at the end after a separation like ------------- . Appending the user's prompt at the end ensures the user's instruction remains fresh in the context window, otherwise it can be forgotten in very long form documents.
* For other types of content, you could use a tool like Pandoc to extract the textual content, then do the same as above to include the user's prompt, so that you offload the conversion work to a 3rd-party tool you don't have to develop. But of course this is going to be prone to a whole host of bugs or documents misreadings...

IMHO text/md file reading would be very useful without being prone to errors nor bugs. But if not implemented, we can already do that just by copy/pasting in the message content.

**Comment by bi4key** [comment]
Maybe this will be helpfull, to find a concept to start RAG.

New one:
https://k33g.hashnode.dev/how-to-make-an-slm-smarter-short-version 

Older one: 
https://k33g.hashnode.dev/how-to-make-a-small-language-model-smarter

**Comment by bi4key** [comment]
Or for RAG simply use this app/method: 
https://github.com/shubham0204/Android-Document-QA

**Comment by Forza130** [comment]
Hi Vali-98, Google has just released its Gemma 3 models with vision. If it's not possible to do Text/PDF documents for now, what about images?

**Comment by Vali-98** [comment]
> what about images?

Currently ChatterUI simply lacks any media storage for both documents and images while the underlying engine, llama.rn lacks VLM capabilities. I do think its possible, but requires a lot of development time.

**Comment by lrq3000** [comment]
For media storage on Android, I have successfully used this library several times on variols projects, it works very well and simplifies a lot the implementation and debugging:

https://github.com/anggrayudi/SimpleStorage

I will be happy to help if you would like to try this route.

But I understand you may want to wait for the engine, llama.rn, to first support documents.

**Comment by Vali-98** [comment]
> But I understand you may want to wait for the engine, llama.rn, to first support documents.

This is probably the main stopper.

Currently, these can be implemented in the app as is:
- Media attachments for chat entries
- Adding state routines for said attachments, updating, deletion etc
- Updating the markdown parser for rendering said attachments based on media (images ,audio etc)

The concern is that the functions above may not be applicable to the actual llama.rn implementation, and it isn't really worth investing time into it unless I take initiative for implementing vision capabilities myself.

**Comment by Vali-98** [comment]
Even though vision capabilities have been added to llama.rn, the initial purpose of this issue (pdf parsing) is something I have decided to put on a 'not-to-implement' list.
